gms/gossiper.cc

```diff
@@ -1602,8 +1598,8 @@ future<> gossiper::real_mark_alive(inet_address addr, endpoint_state& local_stat
         logger.info("InetAddress {} is now UP, status = {}", addr, status);
     }
 
-    co_await _subscribers.for_each([addr, state, pid] (shared_ptr<i_endpoint_state_change_subscriber> subscriber) -> future<> {
-        co_await subscriber->on_alive(addr, state, pid);
+    co_await _subscribers.for_each([&] (shared_ptr<i_endpoint_state_change_subscriber> subscriber) -> future<> {
+        co_await subscriber->on_alive(addr, local_state, permit.id());
```

Comment:
Referring to my earlier comment -- this is an example of a situation that worries me, where the reentrant locking could give us a false sense of security and lead to bugs.

`on_alive` could potentially call gossiper back leading to a modification of the endpoint_state for `addr`, which we would observe in our `local_state` (since it's no longer a copy, but a reference to the map), which would then affect further `on_alive` calls in this loop.

I'm not saying there's a bug now, but allowing this reentrant locking could make it easier for us to write such bugs, we may replace deadlocks which are liveness bugs with safety bugs (breaking invariants etc.).

Reply:
> since it's no longer a copy, but a reference to the map

This is a good point, I'll change this part.
local_state is passed by value to `on_alive` but we need to keep a copy of it across possible preemption.

Reply:
fixed in v2